Corrected lack of XML.ElementInfo text and element child mixing

Spartan322 opened the pull request:
ElementInfo:
	Added XmlNodeType Type
	Added IsText for validating ElementInfo as text element
	Added default ctor that sets Type to Element
	Added argument ctor to streamline construction
	Added TryContentAs methods for bool validation of content
	Added GetContentAs methods for default failed conversions of content
	Added TryAttributeAs methods
	Added GetAttributeAs methods
	Added AttributeAs methods for throwing exceptions
	Added TryAddChild for bool validation of add
	Added TrySetParent for bool validation of parent setting
	Added TrySetAttribute for bool validation attribute setting
	Added TryGetAttribute for bool validation and out of attribute string
	Added SetParent, and SetAttribute for method chaining
	Added GetAttribute for attribute string retrieval with default
	WriteToXML will write text elements with WriteString(Content)
	Added FromText for generating a text element from a string input
Created ElementInfoStringExtensions for string handling
Created ElementInfoListExtensions to take the place of ListExtensions
ListExtensions is obsolete
Created ElementInfoDictionaryExtensions for conversion for the attribute dictionary
EventExecutor implements element and text mixing

Spartan322 committed:
Corrected lack of XML.ElementInfo text and element child mixing
ElementInfo:
	Added XmlNodeType Type
	Added IsText for validating ElementInfo as text element
	Added default ctor that sets Type to Element
	Added argument ctor to streamline construction
	Added TryContentAs methods for bool validation of content
	Added GetContentAs methods for default failed conversions of content
	Added TryAttributeAs methods
	Added GetAttributeAs methods
	Added AttributeAs methods for throwing exceptions
	Added TryAddChild for bool validation of add
	Added TrySetParent for bool validation of parent setting
	Added TrySetAttribute for bool validation attribute setting
	Added TryGetAttribute for bool validation and out of attribute string
	Added SetParent, and SetAttribute for method chaining
	Added GetAttribute for attribute string retrieval with default
	WriteToXML will write text elements with WriteString(Content)
	Added FromText for generating a text element from a string input
Created ElementInfoStringExtensions for string handling
Created ElementInfoListExtensions to take the place of ListExtensions
ListExtensions is obsolete
Created ElementInfoDictionaryExtensions for conversion for the attribute dictionary
EventExecutor implements element and text mixing

Spartan322 committed:
Prevent null attribute dictionary reading from throwing an NRE

Fayti1703 commented:
I'd personally change the names and architecture a little -- specifically, I'd split `ElementInfo` into `NodeInfo` and `ElementInfo : NodeInfo` -- some operations that make sense on elements (e.g. anything related to attributes, having other `ElementInfo`s as children) don't make sense on text nodes.

Spartan322 commented:
> I'd personally change the names and architecture a little -- specifically, I'd split `ElementInfo` into `NodeInfo` and `ElementInfo : NodeInfo` -- some operations that make sense on elements (e.g. anything related to attributes, having other `NodeInfo`s as children) don't make sense on text nodes.

I'm pretty sure that's a breaking change, can be considered for major version change, but I'm fairly sure that would violate semver to do otherwise and it actually could have capability to cause problems for modders if they used 5.1.0 and this is what anything other then 6.0.0 had in it.

Fayti1703 commented:
> > I'd personally change the names and architecture a little -- specifically, I'd split `ElementInfo` into `NodeInfo` and `ElementInfo : NodeInfo` -- some operations that make sense on elements (e.g. anything related to attributes, having other `NodeInfo`s as children) don't make sense on text nodes.
> 
> I'm pretty sure that's a breaking change,

Changing the class architecture (i.e. adding `NodeInfo`) itself shouldn't be, but you're right that changing the method signatures / collection types would be.

I guess for that we could go for something DOMian here -- `Children` that only contains child *elements*, `ChildNodes` that contains all nodes, including text nodes.

Spartan322 commented:
> Something else: The [`As%Type%` methods for attributes](https://github.com/Arkhist/Hacknet-Pathfinder/pull/161/files#diff-22196198ef24bbe256e12ad0c8181b0f660f9d7142de10657318f507eba83847R260) specify the name as _literally_ `attribute[key]`, resulting in `FormatException`s with text like `Value of 'attribute[key]' is not true or false`

This is kind of an intended output, perhaps it should say `Value of 'attribute[key]' is 'not a boolean value'' is not true or false` instead.

Fayti1703 commented:
You sure we don't want, say, `'attribute[foobar]'` in there?

Spartan322 committed:
Make type parse exceptions more useful

AsBoolean, AsFloat, and AsInt FormatExceptions describe both the content value and the key's value for dictionaries

Windows10CE commented on PathfinderAPI/Util/XML/ElementInfo.cs:
```diff
@@ -28,60 +45,222 @@ public override string ToString()
+        => ElementInfoListExtensions.TryGetElement(list, elementName, out info);
+}
+
+public static class ElementInfoDictionaryExtensions
```
Are these not the same extension methods that already exist in PathfinderAPI.Utils.DictionaryExtensions?

Spartan322 replied:
Either I forgot or it was done before we had those, this stuff was around for a long time before I committed it.

Windows10CE commented on PathfinderAPI/Util/XML/ElementInfo.cs:
```diff
@@ -28,60 +45,222 @@ public override string ToString()
+            : throw new FormatException($"Value of '{valName}' is '{content}' which is not a float, e.g.: 1.0");
+}
+
+public static class ElementInfoListExtensions
```
What is being done here? These methods are not extension methods in this class, yet the ones in the obsolete class got their API changed (it was changed from List to IEnumerable)

Spartan322 replied:
This was a screwup, I'll fix it, would've been a breaking change if I don't.

Windows10CE commented on PathfinderAPI/Util/XML/ElementInfo.cs:
```diff
@@ -28,60 +45,222 @@ public override string ToString()
-            : throw new FormatException($"Value of '{Name}' is not a float, e.g.: 1.0");
+        => Content.AsFloat(nameof(Content));
+
+    public bool TryAttributeAsBoolean(string attribName, out bool result)
```
This naming scheme seems confusing, the behavior difference between GetAttributeAsX/AttributeAsX/TryAttributeAsX are not clear and I believe they would be prone to incorrect usage

Spartan322 replied:
Yeah I probably should remove AttributeAs methods and instead add required and default behavior to GetAttributeAs like `GetAttributeAsInt(string name, bool required = false, int def = default)`

Windows10CE commented on PathfinderAPI/Util/XML/ElementInfo.cs:
```diff
@@ -28,60 +45,222 @@ public override string ToString()
+        return Children.Last() == info;
+    }
+
+    public bool TrySetParent(ElementInfo info)
```
I don't think this makes sense to be here, an element doesn't doesn't care about its parent, so why would you set it through the child element

Spartan322 replied:
I suppose, was doing it for convenience but I suppose a child doesn't need to be aware of its parent's add success

Windows10CE commented on PathfinderAPI/Util/XML/ElementInfo.cs:
```diff
@@ -16,6 +19,20 @@ public class ElementInfo
     public List<ElementInfo> Children { get; set; } = new List<ElementInfo>();
     public ulong NodeID { get; } = freeId++;
+    public XmlNodeType Type { get; set; }
+    public bool IsText => Type == XmlNodeType.Text;
```
This entire structure is confusing, why can a text element have attributes and children if they will never be serialized? This also makes adding text to an element two operations for a developer, since I have to both create a child element and also append to Content myself to keep everything up to date. These being out of sync really shouldn't be allowed.

Spartan322 replied:
Backwards compatibility, this was addressed with [here](https://github.com/Arkhist/Hacknet-Pathfinder/pull/161#issuecomment-1110848934). It obviously doesn't make sense but without a refactor which would introduce breaking changes, which is addressed with #163, there is no safe way for mods to rely on the API's semver being true.
